tools/cldr-code/src/main/java/org/unicode/cldr/test/DisplayAndInputProcessor.java

```diff
@@ -1013,4 +998,98 @@ public static NumericType getNumericType(String xpath) {
             return posixDigitCount;
         }
     }
+
+    /**
+     * Turn all whitespace sequences (including tab and newline, and NBSP for certain paths)
+     * into a single space or a single NBSP depending on path.
+     * Also trim initial/final NBSP, unless the value is only the one character, "\u00A0"
+     *
+     * @param path
+     * @param value
+     * @return the normalized value
+     */
+    private String normalizeWhitespace(String path, String value) {
+        PathSpaceType pst = PathSpaceType.get(path);
+        if (pst == PathSpaceType.allowSp) {
+            value = WHITESPACE_AND_NBSP_TO_NORMALIZE.matcher(value).replaceAll(" "); // replace with regular space
+        } else if (pst == PathSpaceType.allowNbsp) {
+            value = WHITESPACE_AND_NBSP_TO_NORMALIZE.matcher(value).replaceAll("\u00A0"); // replace with NBSP
+            value = trimNBSP(value);
+        } else if (pst == PathSpaceType.allowSpOrNbsp) {
+            /*
+             * in this case don't normalize away NBSP
+             */
+            value = WHITESPACE_NO_NBSP_TO_NORMALIZE.matcher(value).replaceAll(" "); // replace with regular space
+            /*
+             * if any NBSP and regular space are adjacent, replace with NBSP
+             */
+            value = NBSP_PLUS_SPACE_TO_NORMALIZE.matcher(value).replaceAll("\u00A0");
+            value = SPACE_PLUS_NBSP_TO_NORMALIZE.matcher(value).replaceAll("\u00A0");
+            value = trimNBSP(value);
+        } else {
+            throw new IllegalArgumentException("Unknown PathSpaceType " + pst);
+        }
+        return value;
+    }
+
+    /**
+     * Delete any initial or final NBSP, unless the value is just NBSP
+     *
+     * @param value
+     * @return the trimmed value
+     */
+    private String trimNBSP(String value) {
+        if (!"\u00A0".equals(value)) {
+            value = INITIAL_NBSP.matcher(value).replaceAll("");
```

Comment:
This also needs changing, I think.

Reply:
Changing how?

Reply:
The unit tests include examples starting and ending with NBSP with or without preceding/following ordinary spaces. If there's a logical error, the best way to show it would be with an example for which this code fails, and then we could add a unit test for it.